src/backend/joanie/core/factories.py

```diff
@@ -221,70 +229,160 @@ def resource_link(self):
+            min_date = self.ref_date
+            max_date = self.start
+        else:
+            # Otherwise, the enrollment start date can be in the past or in the future
+            min_date = self.start - period
+            max_date = self.start
+
         return datetime.utcfromtimestamp(
             random.randrange(  # nosec
-                int((self.start - period).timestamp()), int(self.start.timestamp())
+                int(min_date.timestamp()), int(max_date.timestamp())
             )
         ).replace(tzinfo=timezone.utc)
 
     @factory.lazy_attribute
     def enrollment_end(self):
         """
-        The end of enrollment is a random datetime between the start of enrollment
-        and the end of the course.
-        If the enrollment start and end datetimes have been forced to incoherent dates,
-        then just don't set any end of enrollment...
-        """
-        if not self.start:
+        Compute the enrollment end date according to the course run state
+        and the ref date. Otherwise, pick a random date in the range of enrollment
+        start date and end date.
+        """
+        if (
+            not self.start
+            or not self.enrollment_start
+            or self.state == CourseState.ARCHIVED_OPEN
+        ):
+            # Archived open state is a special case.
+            # The course run has ended but enrollment is still opened.
             return None
-        enrollment_start = self.enrollment_start or self.start - timedelta(
-            days=random.randint(1, 90)  # nosec
-        )
-        max_enrollment_end = self.end or self.start + timedelta(
-            days=random.randint(1, 90)  # nosec
-        )
-        max_enrollment_end = max(
-            enrollment_start + timedelta(hours=1), max_enrollment_end
-        )
+
+        period = timedelta(
+            days=random.randrange(1, 90, 1)  # nosec
+        )  # between 1 and 90 days
+
+        if self.state in [CourseState.ONGOING_OPEN, CourseState.FUTURE_OPEN]:
+            # The course run is opened for enrollment, so the enrollment end date must
+            # be greater than the ref date and less than the course run end
+            min_date = self.ref_date
+            max_date = self.end or self.ref_date + period
+        elif self.state in [CourseState.ONGOING_CLOSED, CourseState.FUTURE_CLOSED]:
+            # The course run is closed for enrollment,
+            # so the enrollment end date must be less than the ref date
+            min_date = self.ref_date - period
+            max_date = self.ref_date
+        elif self.state == CourseState.FUTURE_NOT_YET_OPEN:
+            # The course run is not yet open for enrollment, so the enrollment end date
+            # must be greater than the start date
+            # (as this one is greater than the ref date)
+            min_date = self.enrollment_start
+            max_date = self.end or self.enrollment_start + period
```

Comment:
if end is not None it should be greater than enrollment_start or raise an error (all of these could be forced...)